pkg/abstraction/db.go

```diff
@@ -51,6 +62,164 @@ func (w *dbWrapper) GetValue() interface{} {
 	return w.db.Value
 }
 
+func (w *dbWrapper) getTableName(v reflect.Value) string {
+	t := v.Type()
+
+	// Get TableName, either from type name or if it exists from TableName function
+	nameFunc, exists := t.MethodByName("TableName")
+	if exists {
+		nameResult := nameFunc.Func.Call([]reflect.Value{v})
+		return nameResult[0].String()
+	}
+	return gorm.ToDBName(t.Name())
+}
+
+func (w *dbWrapper) generateWhereClause(query *string, v reflect.Value) error {
+	var primary []string
+
+	t := v.Type()
+
+	// Search for primary keys in the reference type
+	idTag := "ID"
+	_, found := t.FieldByName(idTag)
+	if found {
+		primary = append(primary, idTag)
+	}
+
+	primaryRegExp := regexp.MustCompile("primary_key")
+	for i := 0; i < t.NumField(); i++ {
+		f := t.Field(i)
+		v, ok := f.Tag.Lookup("gorm")
+		if ok {
+			isPrimary := primaryRegExp.MatchString(v)
+			if isPrimary {
+				primary = append(primary, f.Name)
+			}
+		}
+	}
+
+	if len(primary) == 0 {
+		return errors.New("no primary key found")
+	}
+
+	// Set WHERE clause
+	*query = fmt.Sprintf("%s WHERE", *query)
```

Comment:
I found something which could be a potential attack vector but fixed it. Now everything user controlled is checked for sql injections by the sql/db package.